database/ydb/ydb.go

```diff
@@ -0,0 +1,326 @@
+	UPSERT INTO %s (sequence, version, dirty) VALUES ($sequence, $version, $dirty);
+	`, db.config.MigrationsTable)
+
+	if _, err := tx.Exec(query, sql.Named("sequence", time.Now().UnixNano()), sql.Named("version", int64(version)), sql.Named("dirty", dirty)); err != nil {
+		return &database.Error{OrigErr: err, Query: []byte(query)}
+	}
+
+	return tx.Commit()
+}
+
+// migrationTableExists checks if migration table exists
+// returns nil, if table exists
+func (db *YDB) migrationTableExists() error {
+	var (
+		table string
+		query = "SELECT DISTINCT Path FROM `.sys/partition_stats` WHERE Path LIKE '" + db.config.MigrationsTable + "'"
+	)
+
+	res, err := db.conn.QueryContext(ydbsql.WithScanQuery(context.Background()), query)
+	if err != nil {
+		return &database.Error{OrigErr: err, Query: []byte(query)}
+	}
+	if !res.NextResultSet() || !res.Next() {
+		return &database.Error{OrigErr: err, Query: []byte(query)}
+	}
+	if err = res.Scan(&table); err != nil {
+		return &database.Error{OrigErr: err, Query: []byte(query)}
+	}
+
+	return nil
+}
+
+// ensureVersionTable checks if versions table exists and, if not, creates it.
+// Note that this function locks the database, which deviates from the usual
+// convention of "caller locks" in the ClickHouse type.
+func (db *YDB) ensureVersionTable() (err error) {
+	if err = db.Lock(); err != nil {
+		return err
+	}
+
+	defer func() {
+		if e := db.Unlock(); e != nil {
+			if err == nil {
+				err = e
+			} else {
+				err = multierror.Append(err, e)
+			}
+		}
+	}()
+
+	if err := db.migrationTableExists(); err == nil {
+		return nil
+	}
+
+	// if not, create the empty migration table
+	query := fmt.Sprintf(`
+		CREATE TABLE %s (
+			sequence   Int64,
+			version    Int64,
+			dirty      Bool,
+			PRIMARY KEY(sequence)
+		)`, db.config.MigrationsTable)
+
+	if _, err := db.conn.ExecContext(ydbsql.WithSchemeQuery(context.Background()), query); err != nil {
+		return &database.Error{OrigErr: err, Query: []byte(query)}
+	}
+	return nil
+}
+
+func (db *YDB) Drop() (err error) {
+	query := "SELECT DISTINCT Path FROM `.sys/partition_stats` WHERE Path NOT LIKE '%/.sys%'"
+	tables, err := db.conn.QueryContext(ydbsql.WithScanQuery(context.Background()), query)
+
+	if err != nil {
+		return &database.Error{OrigErr: err, Query: []byte(query)}
+	}
+	defer func() {
+		if errClose := tables.Close(); errClose != nil {
+			err = multierror.Append(err, errClose)
+		}
+	}()
+
+	if !tables.NextResultSet() {
+		return nil
+	}
+
+	for tables.Next() {
+		var table string
+		if err := tables.Scan(&table); err != nil {
+			return err
+		}
+
+		query = fmt.Sprintf("DROP TABLE `%s`", table)
+
+		if _, err := db.conn.ExecContext(ydbsql.WithSchemeQuery(context.Background()), query); err != nil {
+			return &database.Error{OrigErr: err, Query: []byte(query)}
+		}
+	}
+
+	if err := tables.Err(); err != nil {
+		return &database.Error{OrigErr: err, Query: []byte(query)}
+	}
+
+	return nil
+}
+
+func (db *YDB) Lock() error {
```

Comment:
Lock must get lock with database (not process) level. For prevent run second migrate process same time on same database from other shell/computer.

https://github.com/golang-migrate/migrate/blob/3061f4564a52e71723da9e0a26e4eb3ba89d0f7b/database/driver.go#L55

Reply:
Until ydb have some locking mechanics in ydb-go-sql this can not be statisfied (but I hope they will).

Reply:
Lock can implemented as updatable record in database with expiration timeout. Can can be doing for any database without special locking mechanics.
For example Mongo: https://github.com/golang-migrate/migrate/blob/4ba695707abdb86fc76501cd96c82abca280dd28/database/mongodb/mongodb.go#L347

Reply:
I'm not a huge fan on hand-made locks on tables. This can be implemented in separate PR if you wish to.

Reply:
It will need additional storage space - for save state.

What about create additional column state now - for use it for locks in next pr.
It will be difficult to change format of version table.

Reply:
I would wait for coordination capabilities to be announced for this.